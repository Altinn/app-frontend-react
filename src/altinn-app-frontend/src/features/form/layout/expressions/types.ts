import type { layoutExpressionFunctions } from 'src/features/form/layout/expressions/functions';

export type LayoutExpressionFunction = keyof typeof layoutExpressionFunctions;

export interface ILayoutExpressionAlias {
  [key: string]: RegExp;
}

export type ILayoutExpressionAliases = {
  [key in LayoutExpressionFunction]: ILayoutExpressionAlias[];
};

export interface ILayoutExpressionDataModelArg {
  dataModel: string;
}

export interface ILayoutExpressionComponentArg {
  component: string;
}

export type ILayoutExpressionArg =
  | string
  | boolean
  | number
  | ILayoutExpressionDataModelArg
  | ILayoutExpressionComponentArg;

export interface ILayoutExpression {
  function: keyof typeof layoutExpressionFunctions;
  args: [ILayoutExpressionArg, ILayoutExpressionArg];
}
