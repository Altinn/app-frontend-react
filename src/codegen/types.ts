import type { CompTypes } from 'src/layout/layout';

export type PropList = {
  [key in CompTypes]: PropObjectProperties;
};

export type PropObjectProperties = {
  [key: string]: ComponentProperty;
};

export type ComponentProperty =
  | PropString
  | PropBool
  | PropNumber
  | PropDate
  | PropAnyExpr
  | PropConst
  | PropLink
  | PropEnum
  | PropUnion
  | PropRange
  | PropObject
  | PropArray;

export interface PropTexts {
  title?: string;
  description?: string;
}

export interface PropBase {
  nb?: PropTexts;
  en?: PropTexts;
  default?: unknown;
  examples?: unknown[];
  deprecated?: boolean;
  required?: boolean;
}

export interface PropMaybeExpression extends PropBase {
  canBeExpression?: boolean;
}

export interface PropString extends PropMaybeExpression {
  type: 'string';
  pattern?: string;
}

export interface PropBool extends PropMaybeExpression {
  type: 'boolean';
}

export interface PropNumber extends PropMaybeExpression {
  type: 'number' | 'integer';
  minimum?: number;
  maximum?: number;
}

export interface PropDate extends PropMaybeExpression {
  type: 'date';
}

export interface PropAnyExpr extends PropMaybeExpression {
  type: 'any';
  canBeExpression: true;
}

export interface PropConst extends PropBase {
  type: 'const';
  const: unknown;
}

export interface PropLink extends Partial<PropBase> {
  type: 'link';
  link: string;
}

export interface PropEnum extends PropBase {
  type: 'enum';
  enum: unknown[];
}

export interface PropUnion extends PropBase {
  type: 'union';
  union: ComponentProperty[];
}

export interface PropRange extends PropBase {
  type: 'range';
  minimum: number;
  maximum: number;
  asText: string;
}

export interface PropObject extends PropBase {
  type: 'object';
  properties: PropObjectProperties;
}

export interface PropArray extends PropBase {
  type: 'array';
  items: ComponentProperty;
  minItems?: number;
  maxItems?: number;
}
