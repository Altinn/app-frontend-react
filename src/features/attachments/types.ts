/**
 * File scan result enum matching the C# backend enum
 */
export type FileScanResult = 'NotApplicable' | 'Pending' | 'Clean' | 'Infected';

/**
 * Attachment processing states for UI feedback
 */
export type AttachmentProcessingState = 'uploading' | 'deleting' | 'updating';

/**
 * Combined attachment states for priority-based messaging
 * Priority order: Infected > uploading/deleting/updating > Pending > ready
 */
export type AttachmentState = FileScanResult | AttachmentProcessingState | 'ready';

/**
 * Return type for enhanced attachment state hook
 */
export interface AttachmentStateInfo {
  hasPending: boolean;
  state: AttachmentState;
}
