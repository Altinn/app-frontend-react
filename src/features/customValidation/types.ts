import type { IExpressionValidationDefinition } from 'src/utils/validation/types';

export type ICustomValidationState = {
  customValidation: IExpressionValidationDefinition | null;
  error: Error | null;
};
