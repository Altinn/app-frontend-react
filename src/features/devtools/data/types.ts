export type DevToolsHiddenComponents = 'show' | 'disabled' | 'hide';

export type IDevToolsState = {
  isOpen: boolean;
  pdfPreview: boolean;
  hiddenComponents: DevToolsHiddenComponents;
  activeTab: DevToolsTab;
  activeSelectionMode: SelectionMode | undefined;
  layoutInspector: {
    selectedComponentId: string | undefined;
  };
  nodeInspector: {
    selectedNodeId: string | undefined;
  };
  dataModelInspector: {
    selectedDataModelBinding: string | undefined;
    selectedPath: string | undefined;
  };
  exprPlayground: {
    expression: string | undefined;
    forPage: string | undefined;
    forComponentId: string | undefined;
  };
  logs: IDevToolsLog[];
};

export type IDevToolsActions = {
  open: () => void;
  close: () => void;
  setActiveTab: (tabName: DevToolsTab) => void;
  setActiveSelectionMode: (mode: SelectionMode | undefined) => void;
  focusLayoutInspector: (componentId: string) => void;
  focusNodeInspector: (nodeId: string) => void;
  focusDataModelInspector: (dataModelBinding: string) => void;
  setPdfPreview: (preview: boolean) => void;
  setShowHiddenComponents: (value: DevToolsHiddenComponents) => void;
  exprPlaygroundSetExpression: (expression: string | undefined) => void;
  exprPlaygroundSetContext: (forPage: string | undefined, forComponentId: string | undefined) => void;
  layoutInspectorSet: (selectedComponentId: string | undefined) => void;
  nodeInspectorSet: (selectedNodeId: string | undefined) => void;
  dataModelInspectorSet: (selectedPath: string | undefined) => void;
  postLogs: (logs: IDevToolsLog[]) => void;
  logsClear: () => void;
};

export type IDevToolsLog = {
  index: number;
  level: 'info' | 'warn' | 'error';
  message: string;
};

export enum DevToolsTab {
  General = 'Generelt',
  Layout = 'Layout',
  Components = 'Komponenter',
  Expressions = 'Uttrykk',
  DataModel = 'Datamodell',
  FeatureToggles = 'Beta-funksjonalitet',
  Logs = 'Logger',
}

export type SelectionMode = 'component' | 'node' | 'dataModel';
