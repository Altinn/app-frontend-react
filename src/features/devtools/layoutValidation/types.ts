import type { lookupBindingInSchema } from 'src/features/datamodel/SimpleSchemaTraversal';
import type { CompIntermediate, CompTypes } from 'src/layout/layout';
import type { LayoutNode } from 'src/utils/layout/LayoutNode';

export interface LayoutValidationCtx<T extends CompTypes> {
  node: LayoutNode<T>;
  item: CompIntermediate<T>;
  lookupBinding(binding: string): ReturnType<typeof lookupBindingInSchema>;
}
