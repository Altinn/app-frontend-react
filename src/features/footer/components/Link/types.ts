import type { IFooterBaseComponent, IFooterIcon } from 'src/features/footer/types';

export interface IFooterLinkComponent extends IFooterBaseComponent<'Link'> {
  title: string;
  target: string;
  icon?: IFooterIcon;
}
