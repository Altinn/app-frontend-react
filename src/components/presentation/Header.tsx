import React from 'react';

import { Grid } from '@material-ui/core';

import classes from 'src/components/presentation/Header.module.css';
import { Progress } from 'src/components/presentation/Progress';
import { useAppSelector } from 'src/hooks/useAppSelector';
import { ProcessTaskType } from 'src/types';
import { getTextFromAppOrDefault } from 'src/utils/textResource';
import type { PresentationType } from 'src/types';

export interface IHeaderProps {
  type: ProcessTaskType | PresentationType;
  header?: string | JSX.Element | JSX.Element[];
  appOwner?: string;
}

export const Header = ({ type, header, appOwner }: IHeaderProps) => {
  const showProgressSettings = useAppSelector((state) => state.formLayout.uiConfig.showProgress);
  const language = useAppSelector((state) => state.language.language);
  const textResources = useAppSelector((state) => state.textResources.resources);

  const showProgress = type !== ProcessTaskType.Archived && showProgressSettings;

  if (!language) {
    return null;
  }

  return (
    <header className={classes.wrapper}>
      <Grid
        container
        direction='row'
        justifyContent='space-between'
        wrap='nowrap'
        spacing={2}
      >
        <Grid item>
          <Grid item>
            <span>{appOwner}</span>
          </Grid>
          <Grid item>
            <h1
              className={classes.headerText}
              data-testid='presentation-heading'
            >
              {type === ProcessTaskType.Archived ? (
                <span>{getTextFromAppOrDefault('receipt.receipt', textResources, language)}</span>
              ) : (
                header
              )}
            </h1>
          </Grid>
        </Grid>
        {showProgress && (
          <Grid
            item
            aria-live='polite'
          >
            <Progress />
          </Grid>
        )}
      </Grid>
    </header>
  );
};
