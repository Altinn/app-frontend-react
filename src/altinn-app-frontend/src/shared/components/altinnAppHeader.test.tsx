import React from 'react';
import { render } from '@testing-library/react';

import Header from './altinnAppHeader';
import { IProfile } from 'altinn-shared/types';

describe('AltinnAppHeader', () => {
  let mockLanguage: any;
  let mockProfile: any;
  beforeEach(() => {
    mockLanguage = {
      language: {
        instantiate: {
          all_forms: 'alle skjema',
          inbox: 'innboks',
          profile: 'profil',
        },
      },
    };
    mockProfile = {
      error: null,
      profile: {
        party: {
          person: {
            firstName: 'Ola',
            middleName: null,
            lastName: 'Privatperson',
          },
          partyId: '123456',
          organisation: null,
        },
      },
    };
  });

  it('should match snapshot', () => {
    const { container } = render(
      <Header
        type='partyChoice'
        language={mockLanguage}
        profile={mockProfile.profile as IProfile}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not render linklist if no type', () => {
    const { container } = render(
      <Header language={mockLanguage} profile={mockProfile} />,
    );

    expect(container.querySelector('ul')).not.toBeInTheDocument();
  });
});
