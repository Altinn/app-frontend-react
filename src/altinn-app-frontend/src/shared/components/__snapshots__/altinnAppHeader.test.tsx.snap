// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AltinnAppHeader should match snapshot 1`] = `
<div
  class="AltinnAppHeader-appBarWrapper-1"
>
  <nav
    class="makeStyles-jumpNav-14"
  >
    <button
      class="makeStyles-button-13 makeStyles-srOnly-15"
      role="link"
    >
      navigation.to_main_content
    </button>
  </nav>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary AltinnAppHeader-default-4 MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular container AltinnAppHeader-toolbarContainer-12 MuiToolbar-gutters"
    >
      <div
        class="MuiGrid-root AltinnAppHeader-logo-9 MuiGrid-item"
      >
        <img
          alt="Altinn logo"
          class="logo"
          id="logo"
          src="https://altinncdn.no/img/Altinn-logo-blue.svg"
        />
      </div>
      <ul
        class="AltinnAppHeader-headerLinkList-6"
      >
        <li
          class="AltinnAppHeader-headerLink-5"
        >
          <a
            href="/ui/Reportee/ChangeReporteeAndRedirect?goTo=/ui/messagebox&R=123456"
          >
            instantiate.inbox
          </a>
        </li>
        <li
          class="AltinnAppHeader-headerLink-5"
        >
          <a
            href="/skjemaoversikt"
          >
            instantiate.all_forms
          </a>
        </li>
        <li
          class="AltinnAppHeader-headerLink-5"
        >
          <a
            href="/ui/Reportee/ChangeReporteeAndRedirect?goTo=/ui/profile&R=123456"
          >
            instantiate.profile
          </a>
        </li>
      </ul>
      <div
        class="a-personSwitcher"
        title="OLA PRIVATPERSON"
      >
        <span
          class="a-personSwitcher-name"
          style="margin-bottom: 10px;"
        >
          <span
            class="d-block"
          />
        </span>
        <i
          aria-hidden="true"
          class="fa fa-private-circle-big AltinnAppHeader-partyIcon-11 AltinnAppHeader-blueDark-2"
        />
      </div>
    </div>
  </header>
</div>
`;
